# Tasks 

## Task 1: Source the Data Set 

Import the Iris data set from the sklearn.datasets module.
Explain, in your own words, what the load_iris() function returns.

In [12]:
# 1. possible way of importing the dataset would be: 
import sklearn as skl
data = skl.datasets.load_iris()
# data 


# alternative method suggested on the sci-kit learn documentation
#from sklearn.datasets import load_iris
#data = load_iris()


The load.iris() function does not return anything, it just loads the data to your dev environment. you have to add the command 'data' to be able to view the dataset. 


 ## Task 2: Explore the Data Structure

In [13]:
data

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  


Based on the official documentation (https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_iris.html), what 'data' returns is a 'bunch', which is a dictionnary-like object with the following elements. in my own words, these are:
- **features** bidimensional array listing the 4 features of each instance in the data set. Each value is a float. 

- **target** a monodimensional array (list) with the target of each instance, expressed in integers (0, 1, 2). 

- **frame** not really sure what this is, will come back to it.

- **target_names** the string (comprehensible) names associated with the target values -- i.e. the name of the flowers (setosa, versicolor, virginica). This is also a bidimensional array, composed of a list of strings (flower names), and a value describing the data type (string: Unicode characters, longest string is less than 10 characters). My reference: https://ds100.org/fa17/assets/notebooks/numpy/Numpy_Review.html. 

- **description** aa short description of the data set, including number of instances, features, classes. 

- **feature_names** monodimensional array: a list including the strings representing the 4 feature names (sepal length and width, petal length and width). 

- **file_name** name of the .csv file 

- **data_module** basically the source, sklearn.datasets. 


## Task 3: Summarize the data

Use numpy to access the data in the dataset and perform the calculations 

In [19]:
# import numpy to work on the features 
import numpy as np

# create variable for features 
features = data.data 
#create variable for target 
target = data.target

print(features)

[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.2]
 [5.  3.2 1.2 0.2]
 [5.5 3.5 1.3 0.2]
 [4.9 3.6 1.4 0.1]
 [4.4 3.  1.3 0.2]
 [5.1 3.4 1.5 0.2]
 [5.  3.5 1.3 0.3]
 [4.5 2.3 1.3 0.3]
 [4.4 3.2 1.3 0.2]
 [5.  3.5 1.6 0.6]
 [5.1 3.8 1.9 0.4]
 [4.8 3.  1.4 0.3]
 [5.1 3.8 1.6 0.2]
 [4.6 3.2 1.4 0.2]
 [5.3 3.7 1.5 0.2]
 [5.  3.3 1.4 0.2]
 [7.  3.2 4.7 1.4]
 [6.4 3.2 4.5 1.5]
 [6.9 3.1 4.

Accessing features in the dataset [to remove]

In [52]:
# create variables to access each feature (= each column in the dataset)
sepal_length = features[:,0]
sepal_width = features[:,1]
petal_length = features[:,2]
petal_width = features[:,3]


**Calculate mean**


In [ ]:
#print(sepal_length.dtype)

# Calculate the mean of each feature in the array. The function takes in the array name and the axis parameter, which allows to specify along which axis the means are computed, in this case column-wise. If no axis is specified, the result is the mean of the flattened array. 
# Sources: https://numpy.org/doc/stable/reference/generated/numpy.mean.html
np.mean(features, axis=0)




array([5.84333333, 3.05733333, 3.758     , 1.19933333])

**Calculate minimum**

In [ ]:
# Use numpy.min to calculate the minimum value of each feature. Source: https://numpy.org/doc/stable/reference/generated/numpy.min.html
# numpy.min works very sismilarly to numpy.mean
np.min(features, axis=0)

array([4.3, 2. , 1. , 0.1])

**Calculate maximum**

In [ ]:
# Use numpy.max to calculate maximum value of each feature. Source: https://numpy.org/doc/stable/reference/generated/numpy.max.html
np.max(features, axis=0)

array([7.9, 4.4, 6.9, 2.5])

**Calculate standard deviation**
 
'The standard deviation is the average amount of variability in your dataset. It tells you, on average, how far each value lies from the mean.'  Pritha Bhandari, How to Calculate Standard Deviation (Guide) | Calculator & Examples, https://www.scribbr.com/statistics/standard-deviation/#:~:text=The%20standard%20deviation%20is%20the,clustered%20close%20to%20the%20mean. 

In this case, petal length is the feature that varies the most across the datasets, with a standard deviation that is two to three times that of the other features. 


In [50]:
# Use numpy.std to calculate standard deviation. Source: https://numpy.org/doc/stable/reference/generated/numpy.std.html
np.std(features, axis=0)

array([0.82530129, 0.43441097, 1.75940407, 0.75969263])

**Calculate median**

Median is defined as the middle term of the given set of data if the data is arranged either in ascending or descending order. Source: https://www.geeksforgeeks.org/median/. 

It differs from mean because it is not necessarily the average of a set of values, only the value that is exactly in the middle. 

In this case, the mean and the median of sepal length and width are similar, while those of petal length and petal width are significantly different. 

Petal length: mean = 3.758; median = 4.35
Petal width: mean = 1.19933333; median: 1.3

In [51]:
# Use numpy.median to calculate median of each feature. Source: https://numpy.org/doc/stable/reference/generated/numpy.median.html
np.median(features, axis=0)

array([5.8 , 3.  , 4.35, 1.3 ])

## Task 4

## Task 5 

## End